### ISSUE TITLE
Config validation

### ISSUE THREAD

**Issue by drubin** [unumotors/mongo-hydra#45]
We are adding more to the config setup. We currently have zero validation on formats and inputs which make the errors incredibly bad when making a mistake.

We should look at adding scheme validation.

Potentially we can use `joi` and document it.

**Comment by pelzerim** [comment]
I agree! This should also happen soon, so we do not have to add it later. Lets add this to the next milestone 0.3

**Comment by pelzerim** [comment]
One thing to also take into account here is the following: CRD support passing of an OpenAPI 3 schema to validate applied schemas (see https://kubernetes.io/blog/2019/06/20/crd-structural-schema/)

This means whatever we do here now (in the CLI stage of this project) should somehow feed into that, so we in the end do not have to write two validations per type.

There are essentially four options:

1. Write the validation as Joi and export that using a library to OpenAPI https://www.npmjs.com/package/joi-openapi
2. Write the validation as Openapi and export that using a library to Joi https://www.npmjs.com/package/openjoi
3. Write the validation as OpenApi and use that to validate https://www.npmjs.com/package/openapi-enforcer
4. Write the validation as Joi and then again write the CRD validation as OpenApi.

I personally think Joi is very easy to write and integrate right now. OpenAPI is more powerful, yes, but it also requires a substantial time investment to use it. I vote 1. We can with the CRD then decide to go with the converter or duplication

**Comment by drubin** [comment]
I also vote for option 1 as its the easiest and provides the most flexibility.

But my reasons are a little different, I believe we are so far away from a CRD implementation that I don't believe the two need to be binary identical. 

For example in the k8s CRD scheme there will be loads of k8s specific configs used to create the replica's and pods/stateful sets which would be totally transparent/opaque to lib/cli implementations. 

For example 

```yaml
replicaSet:
  replicaSetName: mongo-rs
  servers:
  - host: mongo-rs-0:27000
    uri: mongodb://localhost:27000
  - host: mongo-rs-1:27001
    uri: mongodb://localhost:27001
  - host: mongo-rs-2:27002
    uri: mongodb://localhost:27002
```

While this is 100% valid configuration for setting up the replica set it will in no way be enough to create the k8s infrastructure to make this possible. We will need to include things like resource constraints, hard drive type, size,  and lots of other options.

What think will happen is we will use the CRD spec to generate the hydra spec via some sort of mapping logic. For example, the hostname will be calculated based on the stateful set (not user-supplied) 

Also once we have a config  validation setup we can possibly refactor that to use another tool (this should be mostly internal)
